### PR TITLE
Add sheets_create()

### PR DESCRIPTION
**Pull request by jennybc**
HERE WE GO 🚀

My next move is to work on the `sheets` argument .... taking names, taking datasets, etc. But there's already plenty of stuff to decide, so let's get started. I'm going to comment this PR heavily with discussion points.

Below I show a bit of usage.

---

``` r
devtools::load_all("~/rrr/googlesheets4")
#> Loading googlesheets4

sheets_create("sheets-create-demo-1")
#>   Spreadsheet name: sheets-create-demo-1
#>                 ID: 17A4RtHDeEi7L3zBX7FSAXtfhwa-GO3pCalLxCKnLzG4
#>             Locale: en_US
#>          Time zone: Etc/GMT
#>        # of sheets: 1
#> 
#> (Sheet name): (Nominal extent in rows x columns)
#>       Sheet1: 1000 x 26

sheets_create("sheets-create-demo-2", locale = "en_CA")
#>   Spreadsheet name: sheets-create-demo-2
#>                 ID: 1aZqeARCmHXAJcLWmZ5xb4-BhNVHOBxeZLU_tHa3CtEI
#>             Locale: en_CA
#>          Time zone: Etc/GMT
#>        # of sheets: 1
#> 
#> (Sheet name): (Nominal extent in rows x columns)
#>       Sheet1: 1000 x 26

sheets_create(
  "sheets-create-demo-3",
  locale = "fr_FR",
  timeZone = "Europe/Paris"
)
#>   Spreadsheet name: sheets-create-demo-3
#>                 ID: 1LoZnD-swifxuodEDcT4Ta-fVRgKLqVKCOQnTHRLPbgE
#>             Locale: fr_FR
#>          Time zone: Europe/Paris
#>        # of sheets: 1
#> 
#> (Sheet name): (Nominal extent in rows x columns)
#>    Feuille 1: 1000 x 26
```

<sup>Created on 2019-11-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -11,8 +11,8 @@ S3method(ctype,SHEETS_CELL)
 S3method(ctype,character)
 S3method(ctype,default)
 S3method(ctype,list)
-S3method(format,sheets_meta)
-S3method(print,sheets_meta)
+S3method(format,sheets_Spreadsheet)
+S3method(print,sheets_Spreadsheet)
```
Comment (jennybc) [inline NAMESPACE]: This foreshadows an experiment here, which is to start creating companion S3 classes for some of the more import **schemas** for the Sheets API. `Spreadsheet` is an example of that. The existing `sheets_meta` class happened to be exactly this for the `Spreadsheet` schema.

File: R/schema_GridRange.R [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/other#GridRange
```
Comment (jennybc) [inline R/schema_GridRange.R]: There are ~200 so-called schemas in the Sheets API. `GridRange` is an example.

Once you start writing and updating Sheets, you need to create a lot of bodies, in the REST sense, which was not true while we were only reading. And these bodies must be built up of (often nested) schemas. I have to create some sort of machinery around this.

What I do in this PR would not scale nicely to the full set of schemas, but is that necessary? I don't know.

But if I did something more automated re: class creation and validation, then it scales and, in fact, that would belong in gargle, for reuse across multiple APIs.

I don't know what to do.

File: R/schema_NamedRange.R [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  # range is an instance of GridRange
+
+  x
+}
+
+NamedRange <- function(...) {
+  x <- new_NamedRange(...)
+  compact(x)
+}
```
Comment (jennybc) [inline R/schema_NamedRange.R]: I used a few variations on the adv-r S3 regime. Where sometimes I call the validator in the low-level constructor and sometimes in the higher-level constructor (all developer facing). The higher-level constructors generally use `compact()` to get rid of what are often a ton of `NULL` elements.

File: R/schema_NamedRange.R [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  )
+}
+
+validate_NamedRange <- function(x) {
+  # I think read-only vs. required vs. optional status of these elements
+  # depends on what you're trying to do
+  maybe_string(x$namedRangeId, "namedRangeId")
+  maybe_string(x$name, "name")
+
+  # range is an instance of GridRange
```
Comment (jennybc) [inline R/schema_NamedRange.R]: The schema often refer to yet more schema, some of which I haven't implemented. In which case I use this sort of comment in the validator.

Comment (jennybc) [reply R/schema_NamedRange.R]: Although in this case, I have implemented `GridRange`, so I should fix this! But you get the idea.

File: R/schema_SheetProperties.R [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    rightToLeft = rightToLeft
+  )
+  structure(validate_SheetProperties(x), class = "SheetProperties")
+}
+
+validate_SheetProperties <- function(x) {
+  maybe_non_negative_integer(x$sheetId, "sheetId")
+  maybe_string(x$title, "title")
+  maybe_non_negative_integer(x$index, "index")
+  maybe_string(x$sheetType, "sheetType") # enum
```
Comment (jennybc) [inline R/schema_SheetProperties.R]: In addition to ~200 schemas there are *lots* of enums. They aren't particularly easy to pull out of the discovery document in a coherent way. But you could imagine doing so, in a fit of recursive fury, and create a way to declare something a enum, check supplied values against the valid ones, etc.

File: R/schema_Spreadsheet.R [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+Spreadsheet <- function(...) {
+  x <- new_Spreadsheet(...)
+  compact(x)
+}
+
+#  input: instance of Spreadsheet, in the Sheets API sense, as a named list
+# output: instance of sheets_Spreadsheet, which is how I want to hold this info
+sheets_Spreadsheet <- function(x = list()) {
```
Comment (jennybc) [inline R/schema_Spreadsheet.R]: This is the pre-existing function I happened to already have for parsing an instance of `Spreadsheet` into the more natural R object.

File: R/schema_Sheet.R [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  )
+  structure(validate_Sheet(x), class = "Sheet")
+}
+
+validate_Sheet <- function(x) {
+  if (!is.null(x$properties)) {
+    validate_SheetProperties(x)
+  }
+
+  # data is an instance of GridData
```
Comment (jennybc) [inline R/schema_Sheet.R]: I leave this sort of stub for an element I felt moved to include in the schema's S3 class, but that is, itself, a schema I have not bothered to implement.

File: R/schema_SheetProperties.R (hunk 5 above)
Comment (jennybc) [reply R/schema_SheetProperties.R]: Auto-generating an internal data structure and code around the enums sounds much easier to me (maybe even a clear "yes, do this"?) than doing same for the schemas.

**Comment by jennybc** [comment]
OK the constructors for S3 classes based on schemas are very minimalist now.

I must confess I don't see any reason to write these constructors. It feels like I could accomplish the same automatically based on the internal `.schemas` object, i.e. via a hypothetical `new_from_schema(schema)` function. All they do is construct a list with `NULL` elements with specific names. But I will trust in the process.

**Comment by hadley** [comment]
Maybe you shouldn’t then? Maybe it doesn’t make sense to use S3 here at all, and you could just constructor a validation function from the spec?

**Pull request merged**